src/SpectralDataDialog.cpp

```diff
@@ -0,0 +1,251 @@
+   EVT_SHOW(SpectralDataDialog::OnShow)
+   EVT_CLOSE(SpectralDataDialog::OnCloseWindow)
+   EVT_SLIDER(ID_SLIDER_BRUSH_SIZE, SpectralDataDialog::OnBrushSizeSlider)
+END_EVENT_TABLE()
+
+#define Title XO("Spectral Data Control Panel")
+
+SpectralDataDialog::SpectralDataDialog(AudacityProject *parent):
+      wxDialogWrapper(FindProjectFrame( parent ), wxID_ANY, Title,
+                      wxDefaultPosition, wxDefaultSize,
+                      wxDEFAULT_DIALOG_STYLE | wxRESIZE_BORDER )
+{
+   SetName();
+
+   mProject = parent;
+   mSelected = 0;
+   mAudioIOBusy = false;
+
+   //------------------------- Main section --------------------
+   // Construct the GUI.
+   ShuttleGui S(this, eIsCreating);
+   Populate(S);
+
+   wxTheApp->Bind(EVT_AUDIOIO_PLAYBACK,
+                  &SpectralDataDialog::OnAudioIO,
+                  this);
+
+   wxTheApp->Bind(EVT_AUDIOIO_CAPTURE,
+                  &SpectralDataDialog::OnAudioIO,
+                  this);
+
+   Clipboard::Get().Bind(
+         EVT_CLIPBOARD_CHANGE, &SpectralDataDialog::UpdateDisplay, this);
+   parent->Bind(EVT_UNDO_PUSHED, &SpectralDataDialog::UpdateDisplay, this);
+   parent->Bind(EVT_UNDO_MODIFIED, &SpectralDataDialog::UpdateDisplay, this);
+   parent->Bind(EVT_UNDO_OR_REDO, &SpectralDataDialog::UpdateDisplay, this);
+   parent->Bind(EVT_UNDO_RESET, &SpectralDataDialog::UpdateDisplay, this);
+   parent->Bind(EVT_UNDO_PURGE, &SpectralDataDialog::UpdateDisplay, this);
+}
+
+static const AudacityProject::AttachedWindows::RegisteredFactory key{
+      []( AudacityProject &project ){
+         return safenew SpectralDataDialog( &project );
+      }
+};
+
+void SpectralDataDialog::Populate(ShuttleGui & S)
+{
+   auto imageList = std::make_unique<wxImageList>(9, 16);
+   imageList->Add(wxIcon(empty9x16_xpm));
+   imageList->Add(wxIcon(arrow_xpm));
+
+   S.StartVerticalLay(true);
+   {
+      S.StartStatic(XO("Options"), 1);
+      {
+         S.Id(ID_CHECKBOX_OVERTONES)
+               .AddCheckBox(XXO("Enable overtones selection"), false),
+         S.Id(ID_CHECKBOX_SMART)
+               .AddCheckBox(XXO("Enable smart selection"), false);
+      }
+      S.EndStatic();
+
+      S.StartStatic(XO("Brush radius"), 1);
+      {
+         S.Id(ID_SLIDER_BRUSH_SIZE)
+               .Style(wxSL_HORIZONTAL)
+               .Name(XO("Custom brush size"))
+               .AddSlider( {}, 5, 10, 1);
+      }
+      S.EndStatic();
+   }
+   mApplyBtn = S.Id(ID_ON_APPLY)
+         .AddButton(XXO("Apply effect to selection."));
+   S.EndVerticalLay();
+   // ----------------------- End of main section --------------
+
+   Layout();
+   Fit();
+   SetMinSize(GetSize());
+}
+
+void SpectralDataDialog::OnAudioIO(wxCommandEvent& evt)
+{
+   evt.Skip();
+
+   if (evt.GetInt() != 0)
+      mAudioIOBusy = true;
+   else
+      mAudioIOBusy = false;
+}
+
+void SpectralDataDialog::UpdateDisplay(wxEvent& e)
+{
+   e.Skip();
+   if(IsShown())
+      DoUpdate();
```

Comment:
This other bit of event handling copied from the undo history dialog is less useful.  That dialog really needs to detect changes of undo history so it can update what it shows.  This dialog has no such need.